### ISSUE TITLE
[KFNBC]: Last user selection is not remembered

### ISSUE THREAD

**Issue by lugi0** [opendatahub-io/odh-dashboard#303]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

If I go to the spawner and select options which are different from the defaults, and then reload the page, I will be brought back to the default settings

### Expected Behavior

User's last selection always persists across page loads

### Steps To Reproduce

Go to spawner
Select anything that isn't default (e.g. Size from Small to Medium or Minimal Python image)
Reload page

**Comment by DaoDaoNoCode** [comment]
@lugi0 Hi, this design is a little bit different from JupyterHub, we don't record the `lastSelectedImage` and `lastSelectedSize` when the user selects them, because that will cause too many API calls, instead, we only change the value when the user clicks `Start server` button.
Should we change this behavior to make it the same as JupyterHub? @samuelvl

**Comment by lugi0** [comment]
@andrewballantyne I think this also applies to the case in which the user actually spawned the server. 
It is the case for the Tensorflow image at the very least.

**Comment by andrewballantyne** [comment]
Sorry, you're saying you are starting a server and NOT getting your selection remembered @lugi0 ? Because that'd be a bug -- it should remember what you selected on start server button

**Comment by lugi0** [comment]
Actually, I'm seeing inconsistent behaviour. For a couple of times it didn't work, and when I clicked on `Hub Control Panel` from within the JL instance, it would get me to a `Page not found` error page in the dashboard instead of the control panel.
Now I've been trying again with other images (as well as the tensorflow image) and I don't get that Page not found error anymore, and my last selection is remembered.

The only thing I can think of is that an image was being rebuilt while this was happening, but I'm not sure what it could have to do with it

**Comment by lugi0** [comment]
Yes, I can reproduce it, when a build is ongoing the `Hub Control Panel` will redirect to a `Page Not Found` and the last selection won't be remembered
![Screenshot from 2022-08-05 17-20-13](https://user-images.githubusercontent.com/82518733/183109221-aed8c23f-c2e6-45ba-9c89-9a814f60ad93.png)
![Screenshot from 2022-08-05 17-20-56](https://user-images.githubusercontent.com/82518733/183109217-6bb33bd4-7155-4526-bea0-ff0ebe160621.png)
![Screenshot from 2022-08-05 17-21-13](https://user-images.githubusercontent.com/82518733/183109214-f5e99e5a-8958-4d3b-a77c-2e09e557036e.png)

**Comment by lugi0** [comment]
I don't know if you want me to open a new bug for `Hub Control Panel` redirecting to `Page Not Found`, it seems like a separate issue

**Comment by andrewballantyne** [comment]
@lugi0 When you get the 404, can you hold down your back browser button and get the URL to the previous page? I think we keep it in your history when we send you to 404. On Chrome that requires the "show history" and right clicking on the url and "copy link address". 

The url should contain your username (translated if it had invalid k8s name characters). If your username is not in the user state we have, we cannot reform your username and collect all related user data at this time -- so we kick you to 404. 

We should be able to improve this once we keep the notebook around and use it as a source of truth for user data (https://github.com/opendatahub-io/odh-dashboard/issues/387)

It is the same issue likely... we are failing to find your former state. It is currently stored in your ODH Dashboard Config (in the status field) -- feel free check out that status to give anymore information on what data we are using. It is likely a data issue and not a coding issue.

**Comment by lugi0** [comment]
I believe the URL it tries to redirect me to is:
`https://rhods-dashboard-redhat-ods-applications.apps<MY-CLUSTER>.rl7o.s1.devshift.org/notebookController/ldap-2dadmin4/home`

**Comment by lugi0** [comment]
Not sure what to look for in the odhdashboardconfig CRD, this is what I have in the status field:
```
status:
  conditions:
    - type: NamesAccepted
      status: 'True'
      lastTransitionTime: '2022-08-02T11:46:09Z'
      reason: NoConflicts
      message: no conflicts found
    - type: Established
      status: 'True'
      lastTransitionTime: '2022-08-02T11:46:09Z'
      reason: InitialNamesAccepted
      message: the initial names have been accepted
  acceptedNames:
    plural: odhdashboardconfigs
    singular: odhdashboardconfig
    kind: OdhDashboardConfig
    listKind: OdhDashboardConfigList
  storedVersions:
    - v1alpha
```

**Comment by andrewballantyne** [comment]
Sorry, the ODH Dashboard Config IN your dashboard namespace... the CR, not the CRD.

**Comment by lugi0** [comment]
Gotcha, then this is what I have:

```
Status:

  Notebook Controller State:

    Last Selected Image:  

    Last Selected Size:   

    User:                 ldap-user4

    Last Selected Image:  s2i-generic-data-science-notebook:py3.8-1

    Last Selected Size:   Small

    User:                 ldap-user2

    Last Selected Image:  s2i-generic-data-science-notebook:py3.8-1

    Last Selected Size:   Small

    User:                 ldap-user9

    Last Selected Image:  s2i-minimal-notebook:py3.8-1.15.0-qe-nbc-v2

    Last Selected Size:   Small

    User:                 ldap-admin1

    Last Selected Image:  s2i-minimal-notebook:py3.8-1.15.0-qe-nbc-v2

    Last Selected Size:   Small

    User:                 ldap-admin14

    Last Selected Image:  s2i-minimal-notebook:py3.8-1.15.0-qe-nbc-v2

    Last Selected Size:   Small

    User:                 htpasswd-cluster-admin-user

    Last Selected Image:  

    Last Selected Size:   

    User:                 ldap-admin20

    Last Selected Image:  tensorflow:py3.8-cuda-11.4.2-2

    Last Selected Size:   Small

    User:                 

    Last Selected Image:  tensorflow:py3.8-cuda-11.4.2-2

    Last Selected Size:   Small

    User:                 

    Last Selected Image:  s2i-minimal-notebook:py3.8-1.15.0-qe-nbc-v2

    Last Selected Size:   Small

    User:                 

    Last Selected Image:  tensorflow:py3.8-cuda-11.4.2-2

    Last Selected Size:   Small

    User:                 

    Last Selected Image:  tensorflow:py3.8-cuda-11.4.2-2

    Last Selected Size:   Small

    User:                 ldap-admin4

```

**Comment by lugi0** [comment]
I guess the times I got redirected to 404 are those entries without a `User` set?

**Comment by andrewballantyne** [comment]
Correct, we don't know who you are so we can't un-translate the name in the URL back to your state. This is a wonky setup today and should improve when we go to using Notebooks as a source of truth -- because your Notebook will have everything we need to associate you back to it.

**Comment by andrewballantyne** [comment]
For the dev who codes it:

We'll likely implement this via `localStorage` & make use of the [existing hook](https://github.com/opendatahub-io/odh-dashboard/blob/e1bb39fbb385d3b701704c956280676e7ffa4d50/frontend/src/pages/notebookController/screens/server/usePreferredNotebookSize.ts) to pull that information and use it in the Spawner page

**Comment by andrewballantyne** [comment]
As we move forward, this only impacts the initial users who also do not submit the notebook creation. We'll handle what options they have to them via other features as we progress. This issue will become obsolete before we can implement the transient user who doesn't want to submit but wants to store their selection.
